### ISSUE TITLE
aligned struct segfaults garbage collector

### ISSUE THREAD

**Issue by KytoDragon** [ldc-developers/ldc#2235]
I came upon this bug while trying to write a sample program for a different bug (assigning aligned structs in unions overwrites data located after the union).
```
module test.test;

align(4) struct Foo {
    void* y;
    byte z;
}

class Bar {
    
    Foo foo;
    byte x;

    void set(Foo f) {
        x = 99;
        foo = f;
    }
}

void main() {
    Bar bar = new Bar();
    Foo f;
    bar.set(f);
}
```
Stacktrace:
```
Segfault!
Tried to read from 0x0000000000000000
Unknown source: test.exe: _D2gc4impl12conservative2gc3Gcx7prepareMFNbZv
Unknown source: test.exe: _D2gc4impl12conservative2gc3Gcx11fullcollectMFNbbZm
Unknown source: test.exe: _D2gc4impl12conservative2gc14ConservativeGC157__T9runLockedS107_D2gc4impl12conservative2gc14ConservativeGC18fullCollectNoStackMFNbZ2goFNbPS2gc4impl12conservative2gc3GcxZmTPS2gc4impl12conservative2gc3GcxZ9runLockedMFNbKPS2gc4impl12conservative2gc3GcxZm
Unknown source: test.exe: gc_term
Unknown source: test.exe: rt_term
Unknown source: test.exe: _d_run_main
\D\Test_union\__entrypoint.d(8): __entrypoint.main
Unknown source: kernel32.dll: BaseThreadInitThunk
Unknown source: ntdll.dll: RtlUserThreadStart
```
Setting the alignment to 1 or 8 fixes it, setting it to 2 or 4 triggers it. Removing the member x, its assignment in set() or changing the order of members in either Foo or Bar fixes it.
Using Windows 7 x64 using LDC 1.3.0.
This does not happen using DMD 2.075.0

**Comment by KytoDragon** [comment]
Here is my original problem with unions (which does not segfault).
```
module test.test;

align(4) struct Foo {
    void* y;
    byte z;
}

class Bar {
    
    union {
        bool b;
        Foo foo;
    }
    byte x;

    void set(Foo f) {
        x = 99;
        foo = f;
    }
}

void main() {
    Bar bar = new Bar();
    Foo f;
    bar.set(f);
    assert(bar.x == 99); // Fails
}
```

**Comment by kinke** [comment]
> Using Windows 7 x64 using LDC 1.3.0.
> This does not happen using DMD 2.075.0

Yep, most likely because I fixed it upstream but didn't bother cherry-picking it for LDC yet: https://github.com/dlang/dmd/pull/6754

**Comment by kinke** [comment]
Nope, both testcases still fail with that cherry-picked fix. Thanks for reporting; it's apparently an issue wrt. the LLVM IR type layout for the `Foo` struct (LDC assertion: IR type bigger in size than the D type), most likely because we'd need to represent it as a packed IR struct with 3 manual trailing bytes instead of 7 implicit ones. LDC seems not to detect this struct as unnaturally aligned and probably only detects fully packed ones with `align(1)`.

Btw, the correct way of aligning the struct would be:
```D
struct Foo {
    align(4) void* y;
    byte z;
}
```

See the upstream PR for details.

**Comment by kinke** [comment]
Fixed by #2247.
